vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +556,603 @@ void free_route_tree(t_rt_node * rt_node) {
+static void find_sinks_of_route_tree(const t_rt_node* rt_root, vector<int>& sink_nodes) {
+
+	/* Push back the node index of all sinks rooted at rt_root onto sink_nodes */
+
+	t_linked_rt_edge* edge {rt_root->u.child_list};
+	for (;;) {
+		// reached a sink
+		if (!edge) {
+			sink_nodes.push_back(rt_root->inode);
+			return;
+		}
+		// join point (sibling edges)
+		else if (edge->next) {
+			// recursively update for each of its sibling branches
+			do {
+				find_sinks_of_route_tree(edge->child, sink_nodes);
+				edge = edge->next;
+			} while (edge);
+			return;
+		}
+
+		rt_root = edge->child;
+		edge = rt_root->u.child_list;
+	}
+}
+static bool prune_illegal_branches_from_route_tree(t_rt_node* rt_root, float pres_fac, vector<t_rt_node*>& reached_rt_sinks, 
+	vector<int>& remaining_targets, float R_upstream)  {
+
+	/* as we traverse down, do: 											*
+	 *	1. removes illegal branches from pathfinder costs 					*
+	 *	2. populate remaining targets with the illegally routed sinks 		*
+	 *	3. calculate R_upstream and C_downstream for entire subtree 		*
+	 *																		*
+	 *	It is the **caller's responsibility** to actually 					*
+	 *	call pathfinder_update_cost and free_route_subtree 					*
+	 *																		*
+	 *	The removal is only done at branch points for its branches 			*
+	 *	returns whether the entire tree was pruned or not 					*
+	 *	assumes arriving rt_root does not have its R_upstream calculated 	*
+	 *	Tdel for the entire tree after this function is still unset, 		*
+	 *	call load_route_tree_Tdel(rt_root, 0) at SOURCE 					*/
+
+	assert(rt_root != nullptr);
+
+	auto inode = rt_root->inode;
+
+	// illegal tree, propagate information back up (actual removal done at an upstream branch point)
+	if (rr_node[inode].get_occ() > rr_node[inode].get_capacity()) return true;
+	// legal routing, allowed to do calculations now
+
+	// can calculate R_upstream from just upstream information without considering children
+	auto parent_switch = rt_root->parent_switch;
+	if (g_rr_switch_inf[parent_switch].buffered)
+		R_upstream = g_rr_switch_inf[parent_switch].R + rr_node[inode].R;
+	else
+		R_upstream += g_rr_switch_inf[parent_switch].R + rr_node[inode].R;
+	rt_root->R_upstream = R_upstream;
+
+	auto edge = rt_root->u.child_list;
+	// no edge means it must be a sink; legally routed to sink!
+	if (!edge) {
+		reached_rt_sinks.push_back(rt_root);
+		rt_root->C_downstream = 0;	// sink has no node downstream and always has 0 own capacitance
+		return false;
+	}
+	
+	// prune children and get their C_downstream
+	float C_downstream_children {0};
+
+	t_linked_rt_edge* prev_edge {nullptr};
+	do {
+		bool pruned {prune_illegal_branches_from_route_tree(edge->child, pres_fac, reached_rt_sinks, remaining_targets, R_upstream)};
+		// need to remove edge and possibly change first edge if pruned
+		if (pruned) {
+			// optimization for long paths (no sibling edges), propagate illegality up to branch point
```

Comment:
for line 821 & 825 & 838: may change "branch point" to "joint point" for consistency.